Python/context.c

```diff
@@ -119,6 +119,7 @@ static void notify_context_watchers(PyContextEvent event, PyContext *ctx, PyThre
     assert(interp->_initialized);
     uint8_t bits = interp->active_context_watchers;
     int i = 0;
+    PyObject *exc = _PyErr_GetRaisedException(ts);
```

Comment:
Why do this outside the while loop? It means that a callback here cannot trust the exception state -- if they look at the current exception (e.g. using PyErr_GetRaisedException()) they might see the exception left behind by a previous callback.

Reply:
It's done outside the loop to reduce overhead (context switches can be frequent).

Both `_PyErr_GetRaisedException` and `PyErr_FormatUnraisable` clear the error, so each callback should see a null exception, unless a callback raises an exception but returns non-negative.

If this PR is merged I plan on opening another PR that will change the callback return type from `int` to `void` to discourage the use of exceptions.  Instead of checking the return value it will check `_PyErr_Occurred`, guaranteeing that each callback will see a null exception.

Reply:
Why not send a PR that implements your whole plan? Reviewing one bit at a time without knowing where you are ultimately going with this feels counterproductive.

(Also, please don't force push. We already squash merges when eventually merging PRs; squashing while the code is under review makes things more complicated for the reviewers.)

Reply:
See #124776 for everything combined.

I do separate PRs precisely because I saw that the commits are squashed together when merging.  I prefer multiple small commits that each change one thing, so if PRs are squashed then that means multiple PRs.  (Small commits are easier to cherry-pick, easier to roll back if there's a bug, easier to rebase, and most importantly easier to understand.)

If the commits are not squashed, then frequent rebasing and force-pushing is a necessity.

Since I'm trying to keep the PRs limited in scope to a single commit, I'll regular-push fixup commits in the future.  Force-pushing is still required to revise the commit message however.  Also, rebasing is less likely to introduce uncaught errors than merging when resolving merge conflicts with `main`.